File: backend/recipe-service/src/main/java/com/codrut/recipeservice/controller/RecipeController.java
package com.codrut.recipeservice.controller;

import com.codrut.recipeservice.controller.dto.CreateRecipeDTO;
import com.codrut.recipeservice.controller.dto.RecipeDTO;
import com.codrut.recipeservice.controller.dto.ResponseDTO;
import com.codrut.recipeservice.service.RecipeService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import javax.validation.Valid;
import java.util.List;

@RestController
@RequiredArgsConstructor
@RequestMapping("/recipe")
public class RecipeController {

    private final RecipeService service;

    @PostMapping("/")
    public ResponseEntity<ResponseDTO<String>> save(@Valid @RequestBody CreateRecipeDTO dto) {
        service.save(dto);

        return ResponseEntity.ok(new ResponseDTO<>("The recipe was created!"));
    }

    @GetMapping("/")
    public ResponseEntity<ResponseDTO<List<RecipeDTO>>> getAll() {
        return ResponseEntity.ok(new ResponseDTO<>(service.getAll()));
    }

    @GetMapping("/{id}")
    public ResponseEntity<ResponseDTO<RecipeDTO>> getById(@PathVariable Long id) {
        return ResponseEntity.ok(new ResponseDTO<>(service.getById(id)));
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<ResponseDTO<String>> delete(@PathVariable Long id) {
        service.delete(id);

        return ResponseEntity.ok(new ResponseDTO<>("The recipe was deleted!"));
    }

}

File: bakcend/recipe-service/src/main/kotlin/com/codrut/recipe/RecipeController.kt
package com.codrut.recipe


import jakarta.validation.Valid
import jakarta.validation.constraints.NotNull
import jakarta.validation.constraints.Size
import java.util.UUID
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

data class RecipeWriteDto(
    @field:NotNull
    @field:Size(min = 3, max = 32)
    val title: String?,
    @field:NotNull
    @field:Size(min = 3, max = 512)
    val body: String?,
    @field:NotNull
    @field:Size(min = 3, max = 64)
    val createdBy: String?,
) {

    constructor() : this("", "", "")

    fun toEntity(): Recipe = Recipe(
        title = title ?: "",
        body = body ?: "",
        createdBy = createdBy ?: ""
    )
}

@RestController
@RequestMapping("/api/v1/recipe")
class RecipeController(
    private val repo: RecipeRepository
) {

    @PostMapping("/")
    suspend fun createRecipe(@Valid @RequestBody recipeDto: RecipeWriteDto) {
        repo.save(recipeDto.toEntity())
    }

    @GetMapping("/")
    suspend fun getAll() =
        repo.findAll()

    @GetMapping("/{id}")
    suspend fun getById(@PathVariable("id") id: UUID) =
        repo.findById(id)
}